vhost_user_backend/src/lib.rs

```diff
@@ -113,6 +113,40 @@ pub trait VhostUserBackend: Send + Sync + 'static {
     /// A default implementation is provided as we cannot expect all backends
     /// to implement this function.
     fn set_slave_req_fd(&mut self, _vu_req: SlaveFsCacheReq) {}
+
+    /// Get epoll number.
+    /// While multiple queue are defined, it requires the backend to tell how many
+    /// vring handler thread it would like to run. The epoll number is not always
+    /// equal to the queue number, for net, it will be the count of queue pairs.
+    fn epoll_number(&self) -> usize;
+
+    /// Get vring subset.
+    /// While run multiple vring handler thread to service the multiple queues,
+    /// it requires the backend to how many queues each vring handler will handle.
+    fn vring_subset(
+        &self,
+        index: usize,
+        vrings: Vec<Arc<RwLock<Vring>>>,
+    ) -> Vec<Arc<RwLock<Vring>>>;
+
+    /// Get worker thread.
+    /// Get the worker threads from backends if they create internally to handle
+    /// vring events.
+    fn vring_worker_thread(
+        &self,
+    ) -> Option<Vec<thread::JoinHandle<result::Result<(), io::Error>>>> {
+        None
+    }
+
+    /// Tell if worker thread is required to be created
+    /// # Return
+    ///
+    /// * true - Tell the backend lib code to create the vring worker thread
+    /// * false - Tell the backend lib code not to create the vring worker thread
+    ///         since backend has created by itself.
+    fn is_worker_thread_required(&self) -> bool {
+        true
+    }
```

Comment:
What's the reason why we would want the backend library not to create the worker thread? That would defeat the purpose of this crate if the backend implementation was creating the worker thread on its own.

Reply:
Also, this would simplify the code.

Reply:
> What's the reason why we would want the backend library not to create the worker thread? That would defeat the purpose of this crate if the backend implementation was creating the worker thread on its own.


For vhost-user-net backend, the worker threads need to access RxVirtio/TxVirtio and change the buffer there, those data structure defined in backend now, there will be concurrent access which will lead to performance penalty, that's why there is no performance improvement in vhost-user-net backend test although multiple thread is added in vhost-user backend lib code. Thus, each thread should access it's own data structures to avoid race condition, but those structures are vhost-user-net backend specified, it's not reasonable to cover them in vhost-user lib code. To resolve this issue, methods are added to let backends create their own threads.